feature/src/vif.c

```diff
@@ -394,3 +394,159 @@ int vif(int (*read_frame)(float *ref_data, float *main_data, float *temp_data, i
 
     return ret;
 }
+
+int vifdiff(int (*read_frame)(float *ref_data, float *main_data, float *temp_data, int stride, void *user_data), void *user_data, int w, int h, const char *fmt)
+{
+    double score = 0;
+    double scores[4*2];
+    double score_num = 0;
+    double score_den = 0;
+    float *ref_buf = 0;
+    float *ref_diff_buf = 0;
+    float *prev_ref_buf = 0;
+    float *dis_buf = 0;
+    float *dis_diff_buf = 0;
+    float *prev_dis_buf = 0;
+    float *temp_buf = 0;
+    size_t data_sz;
+    int stride;
+    int ret = 1;
+
+    if (w <= 0 || h <= 0 || (size_t)w > ALIGN_FLOOR(INT_MAX) / sizeof(float))
+    {
+        goto fail_or_end;
+    }
+
+    stride = ALIGN_CEIL(w * sizeof(float));
+
+    if ((size_t)h > SIZE_MAX / stride)
+    {
+        goto fail_or_end;
+    }
+
+    data_sz = (size_t)stride * h;
+
+    if (!(ref_buf = aligned_malloc(data_sz, MAX_ALIGN)))
+    {
+        printf("error: aligned_malloc failed for ref_buf.\n");
+        fflush(stdout);
+        goto fail_or_end;
+    }
+    if (!(ref_diff_buf = aligned_malloc(data_sz, MAX_ALIGN)))
+    {
+        printf("error: aligned_malloc failed for ref_diff_buf.\n");
+        fflush(stdout);
+        goto fail_or_end;
+    }
+    if (!(prev_ref_buf = aligned_malloc(data_sz, MAX_ALIGN)))
+    {
+        printf("error: aligned_malloc failed for prev_ref_buf.\n");
+        fflush(stdout);
+        goto fail_or_end;
+    }
+    if (!(dis_buf = aligned_malloc(data_sz, MAX_ALIGN)))
+    {
+        printf("error: aligned_malloc failed for dis_buf.\n");
+        fflush(stdout);
+        goto fail_or_end;
+    }
+    if (!(dis_diff_buf = aligned_malloc(data_sz, MAX_ALIGN)))
+    {
+        printf("error: aligned_malloc failed for dis_diff_buf.\n");
+        fflush(stdout);
+        goto fail_or_end;
+    }
+    if (!(prev_dis_buf = aligned_malloc(data_sz, MAX_ALIGN)))
+    {
+        printf("error: aligned_malloc failed for prev_dis_buf.\n");
+        fflush(stdout);
+        goto fail_or_end;
+    }
+    if (!(temp_buf = aligned_malloc(data_sz * 2, MAX_ALIGN)))
+    {
+        printf("error: aligned_malloc failed for temp_buf.\n");
+        fflush(stdout);
+        goto fail_or_end;
+    }
+
+    int frm_idx = 0;
+    while (1)
+    {
+        ret = read_frame(ref_buf, dis_buf, temp_buf, stride, user_data);
+
+        if(ret == 1){
+            goto fail_or_end;
+        }
+        if (ret == 2)
+        {
+            break;
+        }
+
+        // ===============================================================
+        // offset pixel by OPT_RANGE_PIXEL_OFFSET
+        // ===============================================================
+        offset_image(ref_buf, OPT_RANGE_PIXEL_OFFSET, w, h, stride);
```

Comment:
Not sure if this is the right place to apply this offset. In original VIF, ADM, the offset is -128 is applied before the values are passed into compute_vif or compute_adm to ensure they are 0-mean. Here if you do it before the differencing, it has no effect.

I am also wondering: the output values of the apply_frame_differencing will have range twice of the input values. What would be the consequences when passing the output values to VIF?

Reply:
good question! I think that it might be still ok, since the difference of two zero mean is still zero mean, right?
about the second question, let me think this tomorrow, if it is a problem